zqe/error.go

```diff
@@ -0,0 +1,114 @@
+// Package zqe provides a mechanism to create or wrap errors with information
+// that will aid in reporting them to users and returning them to api callers.
+package zqe
+
+import (
+	"bytes"
+	"fmt"
+	"runtime"
+)
+
+// A Kind represents a class of error. API layers will typically convert
+// these into a domain specific error representation; for example, an http
+// handler can convert these to http specific status codes.
+type Kind int
+
+const (
+	Other Kind = iota
+	Invalid
+	NotFound
+	Exists
+	Conflict
+)
+
+func (k Kind) String() string {
+	switch k {
+	case Other:
+		return "other error"
+	case Invalid:
+		return "invalid operation"
+	case NotFound:
+		return "item does not exist"
+	case Exists:
+		return "item already exists"
+	case Conflict:
+		return "conflict with pending operation"
+	}
+	return "unknown error kind"
+}
+
+type Error struct {
+	Kind Kind
+	Err  error
+}
+
+func pad(b *bytes.Buffer, s string) {
+	if b.Len() == 0 {
+		return
+	}
+	b.WriteString(s)
+}
+
+func (e *Error) Error() string {
+	b := &bytes.Buffer{}
```

Comment:
Nits: Might as well use strings.Builder. And there's no reason to make this a pointer.